src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/Grid.cs

```diff
@@ -285,21 +285,16 @@ public ColumnDefinitionCollection ColumnDefinitions
             }
             set
             {
-                _data ??= new ExtendedData();
-                _data.ColumnDefinitions ??= new ColumnDefinitionCollection(this);
-                _data.ColumnDefinitions.Clear();
-                if (value == null)
+                if (value?.Owner is not null)
                 {
-                    return;
+                    throw new ArgumentException(SR.Format(SR.GridCollection_InOtherCollection, "value", "ColumnDefinitionCollection"));
                 }
-                foreach (ColumnDefinition colDef in value)
+                _data ??= new ExtendedData();
+                _data.ColumnDefinitions?.Clear();
```

Comment:
You can move this inside the `Owner` setter if the value is null. With the `Owner` setter, most of this property could be moved there. Also you don't want to clear someone else's collection.

All you probably need to do is something like this:
```C#
if (value is null)
{
    if (_data?.ColumnDefinitions is { } cd) 
    {
        cd.Owner = null;
        _data.ColumnDefinitions = null; 
    }
}
else
{
    value.Owner = this; // checks for the owner conflict
    _data ??= new ExtendedData()
    _data.ColumnDefinitions = value;
}}

Reply:
Added